consoleme/lib/cloud_credential_authorization_mapping/__init__.py

```diff
@@ -68,6 +74,49 @@ async def retrieve_credential_authorization_mapping(self):
                 return {}
         return self.authorization_mapping
 
+    async def retrieve_reverse_authorization_mapping(self):
+        if (
+            not self.reverse_mapping
+            or int(time.time()) - self.reverse_mapping_last_update > 60
+        ):
+            redis_topic = config.get(
+                "generate_and_store_reverse_authorization_mapping.redis_key",
+                "REVERSE_AUTHORIZATION_MAPPING_V1",
+            )
+            s3_bucket = config.get(
+                "generate_and_store_reverse_authorization_mapping.s3.bucket"
+            )
+            s3_key = config.get(
+                "generate_and_store_reverse_authorization_mapping.s3.file",
+                "reverse_authorization_mapping/reverse_authorization_mapping_v1.json.gz",
+            )
+            try:
+                self.reverse_mapping = await retrieve_json_data_from_redis_or_s3(
+                    redis_topic,
+                    s3_bucket=s3_bucket,
+                    s3_key=s3_key,
+                    json_object_hook=RoleAuthorizationsDecoder,
+                    json_encoder=pydantic_encoder,
+                )
+                self.reverse_mapping_last_update = int(time.time())
+            except Exception as e:
+                sentry_sdk.capture_exception()
+                log.error(
+                    {
+                        "function": f"{self.__class__.__name__}.{sys._getframe().f_code.co_name}",
+                        "error": f"Error loading reverse credential mapping. Returning empty mapping: {e}",
+                    },
+                    exc_info=True,
+                )
+                return {}
+        return self.reverse_mapping
+
+    async def determine_role_authorized_groups(self, account_id: str, role_name: str):
+        arn = f"arn:aws:iam::{account_id}:role/{role_name}"
```

Comment:
Should we force everything to be lowercase when doing the lookup?

Reply:
Good catch! Do we ensure the credential authz mapping has lowercase role ARNs? It looks like they are all lower, but I don't see where that happens after a quick look.

Reply:
Nope! The usernames/groupnames are forced lowercase but the role arns aren't stored lowercase. The loop that checks on authz forces the check to be lowercase. Not a bad optimization to consider

Reply:
Good to know. I'll update the reverse authz mapping generation to lowercase the ARNs.